File: app/schemas/missao.py
"""Pydantic schemas for Missao entity."""

from datetime import date, datetime, time
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import MissaoStatus


class MissaoBase(BaseModel):
    ordem_servico_id: int
    piloto_id: int
    tecnico_id: int | None = None
    drone_id: int
    data_agendada: date
    hora_agendada: time
    area_prevista: Decimal = Field(..., ge=0)
    volume_previsto: Decimal = Field(..., ge=0)
    janela_operacional: str | None = Field(None, max_length=255)
    restricoes: str | None = None
    observacoes_planejamento: str | None = None
    latitude_operacao: Decimal | None = Field(None, ge=-90, le=90)
    longitude_operacao: Decimal | None = Field(None, ge=-180, le=180)
    endereco_operacao: str | None = Field(None, max_length=255)
    referencia_operacao: str | None = Field(None, max_length=255)


class MissaoCreate(MissaoBase):
    pass


class MissaoUpdate(BaseModel):
    piloto_id: int | None = None
    tecnico_id: int | None = None
    drone_id: int | None = None
    data_agendada: date | None = None
    hora_agendada: time | None = None
    area_prevista: Decimal | None = Field(None, ge=0)
    area_realizada: Decimal | None = Field(None, ge=0)
    volume_previsto: Decimal | None = Field(None, ge=0)
    volume_realizado: Decimal | None = Field(None, ge=0)
    janela_operacional: str | None = Field(None, max_length=255)
    restricoes: str | None = None
    observacoes_planejamento: str | None = None
    observacoes_execucao: str | None = None
    latitude_operacao: Decimal | None = Field(None, ge=-90, le=90)
    longitude_operacao: Decimal | None = Field(None, ge=-180, le=180)
    endereco_operacao: str | None = Field(None, max_length=255)
    referencia_operacao: str | None = Field(None, max_length=255)


class MissaoRegistroExecucao(BaseModel):
    area_realizada: Decimal | None = Field(None, ge=0)
    volume_realizado: Decimal | None = Field(None, ge=0)
    observacoes_execucao: str | None = None


class MissaoTransicao(BaseModel):
    status_novo: MissaoStatus
    motivo: str | None = None


class MissaoResponse(BaseModel):
    id: int
    codigo: str
    ordem_servico_id: int
    piloto_id: int
    tecnico_id: int | None
    drone_id: int
    data_agendada: date
    hora_agendada: time
    area_prevista: Decimal
    area_realizada: Decimal | None
    volume_previsto: Decimal
    volume_realizado: Decimal | None
    janela_operacional: str | None
    restricoes: str | None
    observacoes_planejamento: str | None
    observacoes_execucao: str | None
    status: MissaoStatus
    latitude_operacao: Decimal | None
    longitude_operacao: Decimal | None
    endereco_operacao: str | None
    referencia_operacao: str | None
    iniciado_em: datetime | None
    finalizado_em: datetime | None
    encerrado_tecnicamente_em: datetime | None
    encerrado_financeiramente_em: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class HistoricoStatusMissaoResponse(BaseModel):
    id: int
    missao_id: int
    status_anterior: str | None
    status_novo: str
    motivo: str | None
    alterado_por: int
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/missao.ts
/** TypeScript interfaces for Missao entity matching Pydantic schemas. */

import type { MissaoStatus } from './enums';

export interface MissaoResponse {
  id: number;
  codigo: string;
  ordem_servico_id: number;
  piloto_id: number;
  tecnico_id: number | null;
  drone_id: number;
  data_agendada: string;
  hora_agendada: string;
  area_prevista: number;
  area_realizada: number | null;
  volume_previsto: number;
  volume_realizado: number | null;
  janela_operacional: string | null;
  restricoes: string | null;
  observacoes_planejamento: string | null;
  observacoes_execucao: string | null;
  status: MissaoStatus;
  latitude_operacao: number | null;
  longitude_operacao: number | null;
  endereco_operacao: string | null;
  referencia_operacao: string | null;
  iniciado_em: string | null;
  finalizado_em: string | null;
  encerrado_tecnicamente_em: string | null;
  encerrado_financeiramente_em: string | null;
  created_at: string;
  updated_at: string;
}

export interface MissaoCreate {
  ordem_servico_id: number;
  piloto_id: number;
  tecnico_id?: number | null;
  drone_id: number;
  data_agendada: string;
  hora_agendada: string;
  area_prevista: number;
  volume_previsto: number;
  janela_operacional?: string | null;
  restricoes?: string | null;
  observacoes_planejamento?: string | null;
  latitude_operacao?: number | null;
  longitude_operacao?: number | null;
  endereco_operacao?: string | null;
  referencia_operacao?: string | null;
}

export interface MissaoUpdate {
  piloto_id?: number | null;
  tecnico_id?: number | null;
  drone_id?: number | null;
  data_agendada?: string | null;
  hora_agendada?: string | null;
  area_prevista?: number | null;
  area_realizada?: number | null;
  volume_previsto?: number | null;
  volume_realizado?: number | null;
  janela_operacional?: string | null;
  restricoes?: string | null;
  observacoes_planejamento?: string | null;
  observacoes_execucao?: string | null;
  latitude_operacao?: number | null;
  longitude_operacao?: number | null;
  endereco_operacao?: string | null;
  referencia_operacao?: string | null;
}

export interface MissaoRegistroExecucao {
  area_realizada?: number | null;
  volume_realizado?: number | null;
  observacoes_execucao?: string | null;
}

export interface MissaoTransicao {
  status_novo: MissaoStatus;
  motivo?: string | null;
}

export interface HistoricoStatusMissaoResponse {
  id: number;
  missao_id: number;
  status_anterior: string | null;
  status_novo: string;
  motivo: string | null;
  alterado_por: number;
  created_at: string;
}
